### PR TITLE
article for Delegates Practices Fixes #240

### LINKED ISSUE #240: Delegates: Patterns and Practices
This is a sub-task for #63.

Create the article on common patterns and proper practices with delegates.

### PR DESCRIPTION
**Pull request by BillWagner**
This relies on PR #246, which creates the TOC for this section.

### CODE REVIEW

File: docs/languages/csharp/delegates-patterns.md [hunk 1]
```diff
@@ -1,9 +1,323 @@
...
+
 [Previous](delegates-strongly-typed.md)
 
-This article will cover practices used by delegates, and what
-problems are typically solved by creating and using delegates. 
-This address https://github.com/dotnet/core-docs/issues/240
+Delegates provide a mechanism that enables software designs
+involving minimal coupling between compononents.
+
+One excellent examples for this kind of design is LINQ. The LINQ
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: examples-> example

File: docs/languages/csharp/delegates-patterns.md [hunk 2]
```diff
@@ -1,9 +1,323 @@
...
+The only requirement is to provide the implementation of one method that
+is fundamental to the task at hand.
+
+# Building Your Own Components with Delegates
+
+Let's build on that example by creating a component using a design that
+relies on delegates.
+
+Let's define a component that could be used for log messages in a large
+system. The library components could be used in mnany different environments,
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: spelling: many

File: docs/languages/csharp/delegates-patterns.md [hunk 3]
```diff
@@ -1,9 +1,323 @@
...
+on multiple different platforms. There are a lot of common features in the
+component that manages the logs. It will need to accept messages from any
+component in the system. Those message will have different priorities, which
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: message -> messages

File: docs/languages/csharp/delegates-patterns.md [hunk 4]
```diff
@@ -1,9 +1,323 @@
...
+the core component can manage. The messages should have time stamps in their
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: one word: timestamps

File: docs/languages/csharp/delegates-patterns.md [hunk 5]
```diff
@@ -1,9 +1,323 @@
...
+
+There are also combinations of output that might be used in different
+scenarios. You may want to write messages to the console and to a file.
+
+A design based on delegates will provide a great deal of flexibility, and
+make it easy to support storage mechanisms that may be added in the future.
+
+Under this design, the primary log component can be a non-virtual, even
+sealed class. You can plug in any set of delegates to write the messages
+to different storage mediums. The built in support for multicast delegates
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: storage media

File: docs/languages/csharp/delegates-patterns.md [hunk 6]
```diff
@@ -1,9 +1,323 @@
...
+
+```cs
+Logger.WriteMessage -= LogToConsole;
+```
+
+# Practices
+
+Now, you've added a second output handler for the logging sub-system.
+This one needs a bit more infrastructure to correctly support the file
+system. The delgate is an instance method. It's also a private method.
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: delgate -> delegate

File: docs/languages/csharp/delegates-patterns.md [hunk 7]
```diff
@@ -1,9 +1,323 @@
...
+combined with the `Delegate.Invoke()` method:
+
+```cs
+public static void LogMessage(string msg)
+{
+    WriteMessage?.Invoke(msg);
+}
+```
+
+The null conditional operator short-circuits when the left operand
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: maybe add (?.) in parentheses after null conditional operator

File: docs/languages/csharp/delegates-patterns.md [hunk 8]
```diff
@@ -1,9 +1,323 @@
...
+for any writer is to implement one method. That method could be a
+static or instance method. It could be public, private, or any other
+legal access.
+
+The Logger class can make any number of enhancements or changes without
+introducing breaking changes. Like any class, you cannot modify the
+public API without the risk of breaking changes. But, because the
+coupling between the logger and any output engines is only through
+the delegate, no other types (like interfaces or base classes) are
+involved. The coupling is as small as possible.
```
Comment (ghogen) [inline docs/languages/csharp/delegates-patterns.md]: Nice discussion here about loosely coupled components.

File: samples/csharp-language/delegates-and-events/FileLogger.cs [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        try {
+            using (var log = File.AppendText(logPath))
+            {
+                log.WriteLine(msg);
+                log.Flush();
+            }
+        } catch (Exception e)
+        {
+            // Hmm. Not sure what to do.
+            // Logging is failing...
```
Comment (ghogen) [inline samples/csharp-language/delegates-and-events/FileLogger.cs]: Do you want this "not sure what todo" comment in the final version of the code?

Comment (BillWagner) [inline samples/csharp-language/delegates-and-events/FileLogger.cs]: Good point. I updated the comment to explain why there is a catch clause that does nothing.

**Comment by BillWagner** [comment]
responded to edits, rebased and squashed.

**Pull request merged**